JSON fields: logic to guess widget type

elpaso opened the pull request:
Fix #57673 
by trying to examine the first not-null record and check if it's a valid JSON key/value or a JSON array

elpaso committed:
JSON fields: logic to guess widget type

Fix #57673

elpaso committed:
no brace init

rouault commented on src/gui/editorwidgets/qgskeyvaluewidgetfactory.cpp:
```diff
@@ -43,5 +43,41 @@ QgsEditorConfigWidget *QgsKeyValueWidgetFactory::configWidget( QgsVectorLayer *v
+    QgsFeatureRequest req;
+    req.setFlags( Qgis::FeatureRequestFlag::NoGeometry );
+    req.setSubsetOfAttributes( { fieldIdx } );
+    req.setFilterExpression( QStringLiteral( R"("%1" IS NOT NULL)" ).arg( field.name() ) );
```
proper identifier escaping should likely be done for robustness. Same in other files.

I'm also wondering about if doing such feature request could not be sometimes problematic, thinking in particularly to slow datasources like WFS / OGC API Features, where if you are unlucky, it could result in pure client-side evaluation, and ingesting the whole layer if the field appends to be empty.

elpaso replied:
That's a good point. I'll check if there is a way to limit this logic to datasources that don't have this problem.

elpaso replied:
@rouault I couldn't find a capability for that, I changed the logic to NOT do the filtering and just examine the first 20 features (stopping at the first NOT NULL one).

elpaso committed:
Remove the NOT NULL filter and fetch the first 20 features

To avoid ingesting the whole layer in the worst case scenario
of datasources with no filtering capabilities and NULL values.
